src/backends/plonky2/recursion/circuit.rs

```diff
@@ -0,0 +1,627 @@
+/// This file contains the RecursiveCircuit, the circuit used for recursion.
+///
+/// The RecursiveCircuit verifies N proofs of itself (N=arity), together with
+/// the logic defined at the InnerCircuit (in our case, used for the
+/// MainPodCircuit logic).
+///
+/// The arity defines the maximum amount of proofs of itself that the
+/// RecursiveCircuit verifies. When arity>1, using the RecursiveCircuit has the
+/// shape of a tree of the same arity.
+///                                              
+//
+//                      π_root
+//                        ▲
+//                ┌───────┴────────┐
+//                │RecursiveCircuit│
+//                └─▲───▲───▲────▲─┘
+//          ┌───────┘  ┌┘   └┐   └──────┐
+//          │π''_1     │ ... │     π''_N│
+// ┌────────┴───────┐ ┌┴┐┌─┐┌┴┐ ┌───────┴────────┐
+// │RecursiveCircuit│ │.││.││.│ │RecursiveCircuit│
+// └──▲─────────▲───┘ └─┘└─┘└─┘ └──▲─────────▲───┘
+//    │         │                  │         │
+//   π_1  ...  π_N               π'_1 ...  π'_N
+//
+//
+// where
+// N: arity of the RecursiveCircuit
+// π_i: plonky2 proof of the RecursiveCircuit
+//
+use hashbrown::HashMap;
+use plonky2::{
+    self,
+    gates::noop::NoopGate,
+    iop::{
+        target::{BoolTarget, Target},
+        witness::{PartialWitness, WitnessWrite},
+    },
+    plonk::{
+        circuit_builder::CircuitBuilder,
+        circuit_data::{
+            CircuitConfig, CircuitData, ProverCircuitData, VerifierCircuitData,
+            VerifierCircuitTarget,
+        },
+        proof::{ProofWithPublicInputs, ProofWithPublicInputsTarget},
+    },
+    recursion::dummy_circuit::cyclic_base_proof,
+};
+
+use crate::{
+    backends::plonky2::{
+        basetypes::{Proof, C, D},
+        error::{Error, Result},
+    },
+    middleware::F,
+};
+
+/// InnerCircuit is the trait used to define the logic of the circuit that is
+/// computed inside the RecursiveCircuit.
+pub trait InnerCircuit: Clone {
+    type Input;
+
+    fn build(
+        builder: &mut CircuitBuilder<F, D>,
+        // public_inputs will be used for the recursive proof verification
+        public_inputs: Vec<Target>,
+        selectors: Vec<BoolTarget>,
+    ) -> Result<Self>;
+
```

Comment:
I think we need a new method to return the public inputs so that we can pass them from `InnerCircuit` to `RecursiveCircuit`.
For example something like this:
```rust
pub trait InnerCircuit: Clone {
    type InputTarget: Flattenable;

    fn public_inputs(&self) -> Self::InputTarget;
```